packages/compass-generative-ai/src/tools-connection-manager.ts

```diff
@@ -0,0 +1,184 @@
+        "Database Tools in MongoDB Compass makes use of the connection that MongoDB Compass is connected to. \
+To connect, choose a connection from Compass's connection sidebar - https://www.mongodb.com/docs/compass/connect/"
+      )
+    );
+  }
+
+  async connectToCompassConnection(
+    connectParams: ToolsConnectParams
+  ): Promise<AnyConnectionState> {
+    const { connectionId, connectOptions, connectionString } =
+      this.overridePresetAppName(connectParams);
+    try {
+      const serviceProvider = await NodeDriverServiceProvider.connect(
+        connectionString,
+        connectOptions
+      );
+      await serviceProvider.runCommand('admin', { hello: 1 });
+      this.activeConnection = {
+        id: connectionId,
+        provider: serviceProvider,
+      };
+      return this.changeState('connection-success', {
+        tag: 'connected',
+        serviceProvider,
+        isSearchSupported: () => Promise.resolve(false),
+      });
+    } catch (error) {
+      this.logger.error({
+        id: { __value: 1_001_000_411 },
+        context: 'compass-tools-connection-manager',
+        message: `Error connecting to Compass connection - ${
+          error instanceof Error ? error.message : String(error)
+        }`,
+      });
+      return this.changeState('connection-error', {
+        tag: 'errored',
+        errorReason: error instanceof Error ? error.message : String(error),
+      });
+    }
+  }
+
+  override async disconnect(): Promise<ConnectionStateDisconnected> {
+    try {
+      await this.activeConnection?.provider?.close();
+    } catch (error) {
+      this.logger.error({
+        id: { __value: 1_001_000_412 },
+        context: 'compass-tools-connection-manager',
+        message: `Error disconnecting from Compass connection - ${
+          error instanceof Error ? error.message : String(error)
+        }`,
+      });
+    }
+
+    this.activeConnection = null;
+    return this.changeState('connection-close', {
+      tag: 'disconnected',
+    });
+  }
+
+  override async close(): Promise<void> {
+    await this.disconnect();
+    this._events.emit('close', this.currentConnectionState);
+  }
+
+  async updateConnection(
+    connectParams: ToolsConnectParams | undefined
+  ): Promise<void> {
+    if (connectParams?.connectionId === this.activeConnection?.id) {
+      return;
+    }
+
+    await this.disconnect();
+
+    if (!connectParams) {
+      return;
+    }
+
+    if (isAtlasStream(connectParams.connectionString)) {
+      this.logger.warning({
+        id: { __value: 1_001_000_413 },
+        context: 'compass-tools-connection-manager',
+        message: 'Attempting a connection to an Atlas Stream.',
+      });
+      this.changeState('connection-error', {
+        tag: 'errored',
+        errorReason:
+          'Database Tools does not support connecting to Atlas Streams',
+      });
+      return;
+    }
+
+    // TODO: ideally we don't want to wait for the connection to be established
```

Comment:
I read through mongodb-mcp-server's code and if we want to do that we either have to add another layer of connection management here (which is a shame because mcp server already has a connection manager) or do some work in mongodb-mcp-server to be able to "lazy-connect".

ie. start connecting as soon as the user sends the prompt and then by the time the tool call is made it shouldn't matter if it is done connecting.

Ideally, though. Probably something we can compromise on one way or another.